### PR TITLE
UI support for configuring certain mitigations (#1534833)

### PR DESCRIPTION
**Pull request by M4rtinK**
Add the "KERNEL & PERFORMANCE" spoke to GUI and TUI, which makes
it possible to turn off certain security mitigations.

Mitigations that can currently be turned off:

- PTI
- IBRS
- IBPB

The mechanism is similar to (and compatible with) using the
--append option of the bootloader kickstart command.

If a mitigation is turned off via bootloader --append
(for example with "bootloader --append 'nopti') this
will be reflected in the UI. And turning mitigation on and
off will be reflected in the output kickstart.

Resolves: rhbz#1534833

### CODE REVIEW

File: pyanaconda/ui/tui/spokes/performance.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from pyanaconda.flags import flags
+from pyanaconda.i18n import N_, _
+
+from pyanaconda.ui.categories.system import SystemCategory
+from pyanaconda.ui.tui.spokes import EditTUISpoke
+from pyanaconda.ui.tui.spokes import EditTUISpokeEntry as Entry
+
+__all__ = ["PerformanceSpoke"]
+
+class PerformanceSpoke(EditTUISpoke):
```
Comment (poncovka) [inline pyanaconda/ui/tui/spokes/performance.py]: And here.

File: pyanaconda/ui/tui/spokes/performance.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    category = SystemCategory
+
+    edit_fields = [
+        Entry("PTI (Page Table Isolation)", "pti", EditTUISpoke.CHECK, True),
+        Entry("IBRS (Indirect Branch Restricted Speculation)", "ibrs", EditTUISpoke.CHECK, True),
+        Entry("IBPB (Indirect Branch Prediction Barriers)", "ibpb", EditTUISpoke.CHECK, True),
+        ]
+
+    def __init__(self, app, data, storage, payload, instclass):
+        EditTUISpoke.__init__(self, app, data, storage, payload, instclass, "user")
```
Comment (rvykydal) [inline pyanaconda/ui/tui/spokes/performance.py]: Is the "user" okay here?

**Comment by rvykydal** [comment]
Also there are some tests failing.

File: pyanaconda/iutil.py [hunk 3]
```diff
@@ -1288,3 +1291,49 @@ def _wrapper(self, *args, **kwargs):
...
+        opt_set.remove("noibrs")
+
+    # IBPB
+    if no_ibpb:
+        opt_set.add("noibpb")
+    elif "noibpb" in opt_set:
+        opt_set.remove("noibpb")
+
+    # join back to space delimited string and return
+    return " ".join(opt_set)
```
Comment (jkonecny12) [inline pyanaconda/iutil.py]: I really don't think this should be part of the `iutil.py`. It is not used anywhere else than performance spoke so why it is here. We shouldn't use `ituil.py` as code dump for everything. It is hard to find something there that way.

File: pyanaconda/ui/gui/spokes/performance.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        # report that we are done
+        self.initialize_done()
+
+    @property
+    def status(self):
+        # set based on mitigations being turned ON/OFF
+        all_enabled = self._pti_switch.get_active() and \
+                      self._ibrs_switch.get_active() and \
+                      self._ibpb_switch.get_active()
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/performance.py]: `all()` should be used here instead of `and`.

File: pyanaconda/ui/tui/spokes/performance.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        return True
+
+    @property
+    def mandatory(self):
+        return False
+
+    @property
+    def status(self):
+        # set based on mitigations being turned ON/OFF
+        all_enabled = self.args.pti and self.args.ibrs and self.args.ibpb
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/performance.py]: Again use `all` here instead.

File: pyanaconda/ui/tui/spokes/performance.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        # set boot options accordingly
+        opts = iutil.set_bootopt_mitigations(opts=self.data.bootloader.appendLine,
+                                             no_pti=not self.args.pti,
+                                             no_ibrs=not self.args.ibrs,
+                                             no_ibpb=not self.args.ibpb)
+        self.data.bootloader.appendLine = opts
+
+
+
+
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/performance.py]: Remove these additional blank lines.

File: pyanaconda/ui/tui/spokes/performance.py (hunk 2 above)
Comment (M4rtinK) [reply pyanaconda/ui/tui/spokes/performance.py]: Good catch - looks like one can specify the password checking policy like this (kinda strange to have this as part of EditTUISpoke) and the value can be safely left out.

File: pyanaconda/ui/gui/spokes/performance.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+from pyanaconda.ui.gui.spokes import NormalSpoke
+from pyanaconda.ui.categories.system import SystemCategory
+
+import logging
+log = logging.getLogger("anaconda")
+
+__all__ = ["PerformanceSpoke"]
+
+
+class PerformanceSpoke(NormalSpoke):
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/performance.py]: Added.

File: pyanaconda/ui/tui/spokes/performance.py (hunk 1 above)
Comment (M4rtinK) [reply pyanaconda/ui/tui/spokes/performance.py]: Added as well.

File: pyanaconda/ui/gui/spokes/performance.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/performance.py]: Changed.

File: pyanaconda/ui/tui/spokes/performance.py (hunk 5 above)
Comment (M4rtinK) [reply pyanaconda/ui/tui/spokes/performance.py]: Changed.

File: pyanaconda/ui/tui/spokes/performance.py (hunk 6 above)
Comment (M4rtinK) [reply pyanaconda/ui/tui/spokes/performance.py]: Removed.

File: pyanaconda/iutil.py (hunk 3 above)
Comment (M4rtinK) [reply pyanaconda/iutil.py]: Moved to ui/helpers.py - that looks like a good place for it.

File: pyanaconda/ui/gui/spokes/performance.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        self._ibrs_switch.set_active(not disabled_mitigations.no_ibrs)
+        self._ibpb_switch.set_active(not disabled_mitigations.no_ibpb)
+
+        # report that we are done
+        self.initialize_done()
+
+    @property
+    def status(self):
+        # set based on mitigations being turned ON/OFF
+        all_enabled = all(self._pti_switch.get_active(),
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/performance.py]: I think you need to wrap these in array `[]` it won't work this way though.

File: pyanaconda/ui/tui/spokes/performance.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        return True
+
+    @property
+    def mandatory(self):
+        return False
+
+    @property
+    def status(self):
+        # set based on mitigations being turned ON/OFF
+        all_enabled = all(self.args.pti, self.args.ibrs, self.args.ibpb)
```
Comment (jkonecny12) [inline pyanaconda/ui/tui/spokes/performance.py]: Same here.

File: pyanaconda/ui/gui/spokes/performance.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import logging
+log = logging.getLogger("anaconda")
+
+__all__ = ["PerformanceSpoke"]
+
+
+class PerformanceSpoke(NormalSpoke):
+    """ The Kernel & Performance spoke provides tunables for
+    certain security mitigations, that are kernel based and
+    might incur a performance poenalty.
```
Comment (rvykydal) [inline pyanaconda/ui/gui/spokes/performance.py]: if we do one more round, "penoalty" -> "penalty"

File: pyanaconda/ui/tui/spokes/performance.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+from pyanaconda.ui.categories.system import SystemCategory
+from pyanaconda.ui.tui.spokes import EditTUISpoke
+from pyanaconda.ui.tui.spokes import EditTUISpokeEntry as Entry
+
+__all__ = ["PerformanceSpoke"]
+
+class PerformanceSpoke(EditTUISpoke):
+    """ The Kernel & performance spoke provides tunables for
+    certain security mitigations, that are kernel based and
+    might incur a performance poenalty.
```
Comment (rvykydal) [inline pyanaconda/ui/tui/spokes/performance.py]: if we do one more round, "poenalty" -> "penalty"

File: pyanaconda/ui/gui/spokes/performance.py (hunk 8 above)
Comment (M4rtinK) [reply pyanaconda/ui/gui/spokes/performance.py]: Bah, indeed! Should be fixed now.

File: pyanaconda/ui/tui/spokes/performance.py (hunk 9 above)
Comment (M4rtinK) [reply pyanaconda/ui/tui/spokes/performance.py]: Fixed.

**Pull request merged**